opencxl/apps/accelerator.py

```diff
@@ -55,31 +67,269 @@ def __init__(
+                imgbuf.write(cacheline.to_bytes(CACHELINE_LENGTH))
+            im = Image.open(imgbuf)
+
+        return im
+
+    async def _validate_model(self):
+        # pylint: disable=E1101
+        im = await self._get_test_image()
+        tens = cast(torch.Tensor, self.transform(im))
+
+        # Model expects a 4-dimensional tensor
+        tens = torch.unsqueeze(tens, 0)
+
+        pred_logit = self.model(tens)
+        predicted_probs = torch.softmax(pred_logit, dim=1)
+
+        # 10 predicted classes
+        # TODO: avoid magic number usage
+        pred_kv = {self.test_dataset.classes[i]: predicted_probs[i] for i in range(0, 10)}
+
+        json_asenc = str.encode(json.dumps(pred_kv))
+        bytes_size = len(json_asenc)
+
+        json_asint = int.from_bytes(json_asenc)
+
+        RESULTS_HPA = 0x180  # Arbitrarily chosen
+
+        await self._cxl_type1_device.cxl_cache_writelines(RESULTS_HPA, json_asint, bytes_size)
+
+        HOST_VECTOR_ADDR = 0x50
+        HOST_VECTOR_SIZE = 0x58
+
+        await self._cxl_type1_device.cxl_cache_writelines(HOST_VECTOR_ADDR, RESULTS_HPA, 8)
+        await self._cxl_type1_device.cxl_cache_writelines(HOST_VECTOR_SIZE, bytes_size, 8)
+
+        # Done with eval
+        await self._irq_manager.send_irq_request(Irq.ACCEL_VALIDATION_FINISHED)
+
+    async def _run_app(self):
+        # pylint: disable=unused-variable
+        # pylint: disable=E1101
+        logger.info(
+            self._create_message(f"Changing into accelerator directory: {self.accel_dirname}")
+        )
+        os.chdir(self.accel_dirname)
+
+        logger.info(self._create_message("Creating symlinks to training and validation datasets"))
+        os.symlink(src="../imagenette2-160/train", dst="train", target_is_directory=True)
+        os.symlink(src="../imagenette2-160/val", dst="val", target_is_directory=True)
+
+        logger.info(self._create_message("Beginning training"))
+        if torch.cuda.is_available():
+            device = torch.device("cuda:0")
+        elif torch.backends.mps.is_available():
+            device = torch.device("mps")
+        else:
+            device = torch.device("cpu")
+        print(f"torch.device: {device}")
+
+        # Uses CXL.cache to copy metadata from host cached memory into device local memory
+        await self._get_metadata()
+
+        epochs = 2
+        epoch_loss = 0
+        for epoch in range(epochs):
+            loss_fn = torch.nn.CrossEntropyLoss()
+            optimizer = torch.optim.SGD(self.model.parameters())
+            scheduler = torch.optim.lr_scheduler.LinearLR(
+                optimizer, start_factor=1, end_factor=0.5, total_iters=30
+            )
+            self._train_one_epoch(
+                train_dataloader=self.train_dataloader,
+                test_dataloader=self.test_dataloader,
+                optimizer=optimizer,
+                loss_fn=loss_fn,
+                device=device,
+            )
```

Comment:
No assignment for `self.train_dataloader` (and test). Is this supposed to be skeleton code only?

Btw, let's call it `self._train_dataloader`.

Reply:
The assignment for self.train_dataloader was commented out as it relies on a file that doesn't exist yet. As for the rename -- sure!